src/Compilers/CSharp/Portable/Syntax/NonRecursiveSyntaxRewriter.cs

```diff
@@ -0,0 +1,286 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// Represents a non-recursive visitor which descends an entire <see cref="CSharpSyntaxNode"/> graph and
+    /// may replace or remove visited SyntaxNodes in depth-first order.
+    /// </summary>
+    public partial class NonRecursiveSyntaxRewriter : CSharpSyntaxVisitor<NonRecursiveSyntaxRewriter.Chunk>
```

Comment:
>NonRecursiveSyntaxRewriter [](start = 25, length = 26)

The recursive rewriter allows the developer to know the state before and after the child elements are rewritten.  If I understand your implementation correctly, the developer will only know the after state.  It would be useful to be able to access the original node/token when deciding on how to rewrite the node itself.  Possibly this could be a property.